torch/_higher_order_ops/while_loop.py

```diff
@@ -0,0 +1,126 @@
+import torch.utils._pytree as pytree
+
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import (
+    _has_potential_branch_input_alias,
+    _has_potential_branch_input_mutation,
+    _maybe_run_with_interpreter,
+    autograd_not_implemented,
+    UnsupportedAliasMutationException,
+)
+from torch._ops import HigherOrderOperator
+
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    make_fx,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+
+
+while_loop_op = HigherOrderOperator("while_loop")
```

Comment:
We should probably subclass HigherOrderOp and override the `__call__`. The call should be responsible for (1) checking the types and (2) calling `super().__call__`.

We want to make sure:
1. cond_fn is a FX graph
2. body_fn is a FX graph
3. operands is a Tuple.

Yes, that means we'll probably need to change our tests to do a make_fx on cond_fn/body_fn first before calling torch.ops.higher_order.while_loop.

These invariants make it clearer what the torch.ops.higher_order.while_loop supports (FX graphs) vs what torch.while_loop (arbitrary Python functions) supports.

Reply:
Also we can check that operators is a flat tuple of Tensors. Do we need a torch.ops.higher_order.while_loop vs torch.ops.higher_order.while_loop_impl separation, or is just one fine?

Reply:
We only need torch.ops.higher_order.while_loop. It's also to keep it consistent with torch.cond vs torch.ops.higher_order.cond.